db/db_impl/db_impl_write.cc

```diff
@@ -1869,27 +1868,26 @@ Status DBImpl::DelayWrite(uint64_t num_bytes, WriteThread& write_thread,
   return s;
 }
 
-// REQUIRES: mutex_ is held
 // REQUIRES: this thread is currently at the front of the writer queue
-void DBImpl::WriteBufferManagerStallWrites() {
-  mutex_.AssertHeld();
-  // First block future writer threads who want to add themselves to the queue
-  // of WriteThread.
-  write_thread_.BeginWriteStall();
-  mutex_.Unlock();
-
-  // Change the state to State::Blocked.
-  static_cast<WBMStallInterface*>(wbm_stall_.get())
-      ->SetState(WBMStallInterface::State::BLOCKED);
-  // Then WriteBufferManager will add DB instance to its queue
-  // and block this thread by calling WBMStallInterface::Block().
-  write_buffer_manager_->BeginWriteStall(wbm_stall_.get());
-  wbm_stall_->Block();
-
-  mutex_.Lock();
-  // Stall has ended. Signal writer threads so that they can add
-  // themselves to the WriteThread queue for writes.
-  write_thread_.EndWriteStall();
+void DBImpl::MaybeWriteBufferManagerStallWrites() {
+  bool begin_stall =
+      write_buffer_manager_->MaybeBeginWriteStall(wbm_stall_.get());
+  if (begin_stall) {
+    mutex_.Lock();
+    // Block future writer threads who want to add themselves to the queue
+    // of WriteThread.
+    write_thread_.BeginWriteStall();
```

Comment:
Optimization 1: 

Block future writers only if `begin_stall` is true here. That is because we could have runtime-changed the parameters (e.g, allow_stall_, buffer_size_) that affects stall conditions between `if (UNLIKELY(status.ok() && write_buffer_manager_->ShouldStall())) {` and  `bool begin_stall = write_buffer_manager_->MaybeBeginWriteStall(wbm_stall_.get());`